axlearn/common/checkpointer_orbax_emergency_replicator.py

```diff
@@ -0,0 +1,323 @@
+    # Get process ID and IP of jax coordinator
+    process_id, coordinator_address = _retrieve_jax_init_info(parsed_args["local_ckpt_dir"])
+    FLAGS.process_id = int(process_id)
+    FLAGS.distributed_coordinator = coordinator_address
+    FLAGS.experimental_orbax_use_distributed_process_id = True
+
+    yield
+
+
+def _wait_for_file_to_disappear(f, timeout=300):
+    for _ in range(timeout):
+        if not f.exists():
+            return True
+        time.sleep(1)
+    logging.error("File %s did not dissappear in time.", f)
+    return False
+
+
+def _extract_step(f):
+    # The base file name is formatted as {job_name}-s{step}-n{node_rank}-w{worker_rank}
+    return f.rsplit("-", 3)[1][1:]
+
+
+def _block_and_process_restore_dir(directory, timeout=300):
+    """Block until the directory symlink ending with `.restore` appears, then extract
+    the step number and rename the directory using the step number.
+    """
+    suffix = ".restore"
+    for _ in range(timeout):
+        files = os.listdir(directory)
+        for f in files:
+            if f.endswith(suffix):
+                step = _extract_step(f)
+                if step != "0":
+                    os.rename(epath.Path(directory) / f, epath.Path(directory) / step)
+                    logging.info(
+                        "Renamed restore directory at step %s to %s.",
+                        step,
+                        epath.Path(directory) / step,
+                    )
+                else:
+                    logging.info("Found a restore directory at step 0, skipping renaming.")
+                return
+        time.sleep(1)
+    logging.info("%d seconds have passed but no .restore file was found.", timeout)
+
+
+def _retrieve_jax_init_info(local_ckpt_dir):
+    """Retrieve JAX init info from a local file."""
+    jax_init_info_file = "jax-init-info.txt"
+    local_jax_init_info_file = epath.Path(local_ckpt_dir) / jax_init_info_file
+    # Allow time for the JAX init info file to be populated by GKE.
+    # File only populated when the worker with process id of 0 is determined.
+    for i in range(900):
+        if local_jax_init_info_file.exists():
+            return local_jax_init_info_file.read_text().split("\n")[:2]
+        logging.info(
+            "Unable to locate %s after %d seconds, sleeping for 1 second before retrying...",
+            jax_init_info_file,
+            i,
+        )
+        time.sleep(1)
+    raise RuntimeError(
+        f"Unable to locate {jax_init_info_file} after 900 seconds, "
+        "returning empty process id and coordinator address."
+    )
+
+
+class OrbaxEmergencyReplicatorCheckpointer(BaseCheckpointer):
+    """Checkpointer implementation that uses Orbax emergency replicator checkpointer.
+
+    EXPERIMENTAL. Do not use for actual training runs since the checkpoint layout will likely
+    change in the future."""
+
+    @config_class
+    class Config(BaseCheckpointer.Config):
```

Comment:
I'm guessing this is the class that will have fields for how often to checkpoint and backup?